### PR TITLE
Move to a watcher model for controllers

### PR DESCRIPTION
**Pull request by stuartwdouglas**
This is still not perfect as it is not clear how to handle contaminants, however this does remove the pipeline run and task run controllers.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -22,7 +22,10 @@ vet: ## Run go vet against code.
 	go vet ./cmd/... ./pkg/...
 
 test: fmt vet envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -v ./pkg/...
+	go test -v ./pkg/...
+
+e2etest: fmt vet envtest ## Run tests.
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -v ./pkg/... ./test/...
```
Comment (gabemontero) [inline Makefile:28]: this was technically running both the e2e's and unit tests.
```suggestion
	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -v ./test/...
```

File: pkg/apis/jvmbuildservice/v1alpha1/artifactbuildrequest_types.go [hunk 2]
```diff
@@ -27,8 +26,6 @@ const (
 	ArtifactBuildRequestStateNew = "ArtifactBuildRequestNew"
 	// ArtifactBuildRequestStateDiscovering The discovery pipeline is running to try and figure out how to build this artifact
 	ArtifactBuildRequestStateDiscovering = "ArtifactBuildRequestDiscovering"
-	// ArtifactBuildRequestStateDiscovered The discovery process is complete
-	ArtifactBuildRequestStateDiscovered = "ArtifactBuildRequestDiscovered"
```
Comment (gabemontero) [inline pkg/apis/jvmbuildservice/v1alpha1/artifactbuildrequest_types.go:31]: ok this lines up with what i recall from my avoid multiple status writers POC, so I take this as confirming / good

Comment (gabemontero) [inline pkg/apis/jvmbuildservice/v1alpha1/artifactbuildrequest_types.go:31]: ok this lines up with what i recall from my avoid multiple status writers POC, so I take this as confirming / good

File: pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go [hunk 3]
```diff
@@ -59,43 +67,91 @@ func (r *ReconcileArtifactBuildRequest) Reconcile(ctx context.Context, request r
...
+	//there should only be one, be guard against multiple
+	for _, current := range trl.Items {
+		if tr == nil || tr.CreationTimestamp.Before(&current.CreationTimestamp) {
+			tr = &current
+		}
+	}
+	if tr == nil {
+		//no pipeline, very odd
+		//move back to new
+		abr.Status.State = v1alpha1.ArtifactBuildRequestStateNew
```
Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:126]: can you put a TODO comment to remind me to generate an event when this happens

I'll handle dealing with events in a separate PR

Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:126]: can you put a TODO comment to remind me to generate an event when this happens

I'll handle dealing with events in a separate PR

File: pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go [hunk 4]
```diff
@@ -59,43 +67,91 @@ func (r *ReconcileArtifactBuildRequest) Reconcile(ctx context.Context, request r
...
+		}
+	}
+
 	//now let's create the dependency build object
 	//once this object has been created its resolver takes over
 	if abr.Status.Tag == "" {
 		//this is a failure
 		abr.Status.State = v1alpha1.ArtifactBuildRequestStateMissing
-		return reconcile.Result{}, r.client.Status().Update(ctx, &abr)
+		return reconcile.Result{}, r.client.Status().Update(ctx, abr)
```
Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:154]: can you put a TODO comment to remind me to generate an event when this happens

I'll handle dealing with events in a separate PR

Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:154]: can you put a TODO comment to remind me to generate an event when this happens

I'll handle dealing with events in a separate PR

File: pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go [hunk 5]
```diff
@@ -137,7 +196,13 @@ func (r *ReconcileArtifactBuildRequest) handleStateDiscovered(ctx context.Contex
 		}
 	} else {
 		//build already exists, add us to the owner references
+		var recent *v1alpha1.DependencyBuild
 		for _, db := range list.Items {
+			//there should only ever be one build
+			//if somehow we do end up with two we consider the most recent one
+			if recent == nil || recent.CreationTimestamp.Before(&db.CreationTimestamp) {
```
Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:203]: I wonder it this situation would also warrant an event for debug purposes

let's put a TODO comment to at least think about logging or posting an event

Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:203]: I wonder it this situation would also warrant an event for debug purposes

let's put a TODO comment to at least think about logging or posting an event

File: pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go [hunk 6]
```diff
@@ -196,6 +266,60 @@ func (r *ReconcileArtifactBuildRequest) handleStateComplete(ctx context.Context,
...
+		Namespace:     abr.Namespace,
+		LabelSelector: labels.SelectorFromSet(lbls),
+	}
+
+	if err := r.client.List(ctx, list, listOpts); err != nil {
+		return reconcile.Result{}, err
+	}
+	if len(list.Items) == 0 {
+		//we don't have a build for this ABR, this is very odd
+		//move back to new and start again
```
Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:284]: another event TODO comment

Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:284]: another event TODO comment

File: pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go [hunk 7]
```diff
@@ -196,6 +266,60 @@ func (r *ReconcileArtifactBuildRequest) handleStateComplete(ctx context.Context,
...
+		abr.Status.State = v1alpha1.ArtifactBuildRequestStateNew
+		return reconcile.Result{}, r.client.Status().Update(ctx, abr)
+	}
+
+	//let's see if the build has completed
+	var recent *v1alpha1.DependencyBuild
+	for _, db := range list.Items {
+		//there should only ever be one build
+		//if somehow we do end up with two we consider the most recent one
+		if recent == nil || recent.CreationTimestamp.Before(&db.CreationTimestamp) {
```
Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:294]: again with the TODO to consider an event

Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:294]: again with the TODO to consider an event

File: pkg/reconciler/artifactbuildrequest/controller.go [hunk 8]
```diff
@@ -1,31 +1,18 @@
...
-			GenericFunc: func(e event.GenericEvent) bool {
-				//TODO possibly change to false over time
-				return true
-			},
-		})).Complete(r)
+	return ctrl.NewControllerManagedBy(mgr).For(&v1alpha1.ArtifactBuildRequest{}).
+		//we can't use .Owns() here as controller is not true in the owner ref
+		Watches(&source.Kind{Type: &v1beta1.TaskRun{}}, &handler.EnqueueRequestForOwner{OwnerType: &v1alpha1.ArtifactBuildRequest{}, IsController: false}).
+		Watches(&source.Kind{Type: &v1alpha1.DependencyBuild{}}, &handler.EnqueueRequestForOwner{OwnerType: &v1alpha1.ArtifactBuildRequest{}, IsController: false}).
+		Complete(r)
```
Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/controller.go:17]: before reading the rest of this comment, to be clear, I like this ^^ more :-)

but to complete the loop on our conversation this morning with @sbose78 and the "shipwright" form for doing something similar that I have references a bunch of times recently, just so you are aware more precisely of what I'm referring to, where you are clearly well versed now in controller runtime nuance to compare/contrast, here it is:  https://github.com/shipwright-io/build/blob/main/pkg/reconciler/buildrun/controller.go 

take a peek as your curiosity deems worthy @stuartwdouglas

Comment (gabemontero) [inline pkg/reconciler/artifactbuildrequest/controller.go:17]: before reading the rest of this comment, to be clear, I like this ^^ more :-)

but to complete the loop on our conversation this morning with @sbose78 and the "shipwright" form for doing something similar that I have references a bunch of times recently, just so you are aware more precisely of what I'm referring to, where you are clearly well versed now in controller runtime nuance to compare/contrast, here it is:  https://github.com/shipwright-io/build/blob/main/pkg/reconciler/buildrun/controller.go 

take a peek as your curiosity deems worthy @stuartwdouglas

**Review by gabemontero — COMMENTED** [review]
just some minor items, mostly adding TODO's for some future debug enhancements @stuartwdouglas 

definitely like where our "hybrid" unit tests are going

do you think the unit tests are more or less done for now?  I am thinking "yes", but to be honest, it will be easier to tell after this merges and I can run them, look at code coverage, etc.

if there are some major items you know of, either open a github issue here or put a comment in our active jira so I can add them 

otherwise, after this merges, I can dive into more of the e2e test stuff, both here in repo and the separate test repo the team has

File: pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go (hunk 3 above)
Comment (stuartwdouglas) [reply pkg/reconciler/artifactbuildrequest/artifactbuildrequest.go:126]: Shouldn't every state transition trigger an event?

**Pull request merged**